### PR TITLE
[WIP] Flattening algorithm

### PR DESCRIPTION
**Pull request by Vilin97**
EDIT: introducing the flattening algorithm: turning all diffusive hops into reactions and using a standard SSA to simulate.

OLD:
This PR is the first step towards simulating spatially inhomogeneous systems. 
The main contribution is the function `to_spatial_jump_prob` that takes in a graph, diffusion rates and the parameters of a non-spatial problem and outputs a spatial problem, with the number of species equal (number of nodes)*(number of non-spatial species), and diffusive hops are represented as reactions. For example, a diffusive hop of molecule A from node i to node j is represented via the reaction A_i --> A_j. A more detailed description of the function is given in https://github.com/SciML/DiffEqJump.jl/issues/130. 

Another contribution is a function to animate the solution of a 2d network on a grid, which outputs an animation, which can then be saved as a .gif file.

Additionally, there are three examples of the usage of the newly added functionality: `ABC.jl`, `SIR.jl` and `gene_expression.jl`. 
`ABC.jl` sets up a simulation of the network A+B <--> C on a grid (2d or 3d), and does plotting, animation and benchmarking. 
`SIR.jl` shows the flexibility of the script by setting up and solving the SIR network from https://discourse.julialang.org/t/lightgraph-modeltoolkit-optimization/40696. In particular, the network in `SIR.jl` involves a reaction of form S_i + I_j --> 2I_j, where i and j are neighboring nodes. This functionality is achieved by passing in a function `assign_products` into `to_spatial_jump_prob`. 
Finally, `gene_expression.jl` shows how to take an existing non-spatial network, supply an arbitrary graph (a cycle graph C_n is used now) and obtain the spatial variant of the problem. 

While visually the script works correctly, and some testing has been done on the three models above, I have not done extensive testing.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -11,6 +11,7 @@ DiffEqBase = "2b5f629d-d688-5b77-993f-72d75c75574e"
 FunctionWrappers = "069b7b12-0de2-55c6-9aab-29f3d0a68a2e"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 Parameters = "d96e819e-fc66-5662-9728-84c9c7592b0a"
+Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
```
Comment (isaacsas) [inline Project.toml:14]: Let's not add a Plots dependency -- that is quite heavy and could slow package loading quite a bit. 

I'd suggest moving the plotting utils to the tests/spatial folder for now. If we want to create simple plotting routines we can move those to a separate repo down the line.

File: src/spatial/utils.jl [hunk 2]
```diff
@@ -0,0 +1,85 @@
+using Plots
```
Comment (isaacsas) [inline src/spatial/utils.jl:1]: remove

File: src/spatial/to_spatial.jl [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+(optional keyword arg) diff_rates -- diffusion rates (analagous to reaction rates). 0.0 by default.
+(optional keyword arg) save_positions -- when/whether to save positions. Equal to (false, false) by default.
+(optional keyword arg) starting_state -- note the ordering of species. Fills all nodes with prob.u0 by default.
+(optional keyword arg) get_rate -- function to get the rate of a reaction between neighboring nodes. Equal to (rx, _, _, rates) -> rates[rx] by defualt
+(optional keyword arg) assign_products -- function to decide where products are assigned
+
+The ordering of species is:
+[species in node 1, species in node 2, ... ] <-- lengths = num_species, num_species, ...
+"""
+function JumpProblem(prob, aggregator::WellMixedSpatial, massaction_jump::MassActionJump; connectivity_list = nothing, diff_rates = 0.0, save_positions = typeof(prob) <: DiffEqBase.AbstractDiscreteProblem ? (false,false) : (true,true), rng = Xorshifts.Xoroshiro128Star(rand(UInt64)), starting_state = vcat([prob.u0 for i in 1:length(connectivity_list)]...), get_rate = (rx, _, _, rates) -> rates[rx], assign_products = nothing, kwargs...)
```
Comment (isaacsas) [inline src/spatial/to_spatial.jl:20]: Why not allow `ConstantRateJump`s too?

Comment (isaacsas) [reply src/spatial/to_spatial.jl:20]: Can you try to keep lines to ~80 characters so this is readable?

File: src/spatial/to_spatial.jl [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+    if connectivity_list === nothing
+        error("No graph was given. Constructing a non-spatial problem.")
```
Comment (isaacsas) [inline src/spatial/to_spatial.jl:22]: "Construct a non-spatial problems."?

**Comment by isaacsas** [comment]
Something has gotten really messed up here. It looks like you didn't merge master into the PR right. You might want to try that again. Gitkraken is a nice GUI you can use to handling merge conflicts.

**Comment by Vilin97** [comment]
I think I just didn't delete the merge indicators. Should be good now.

File: Project.toml [hunk 5]
```diff
@@ -10,6 +10,7 @@ DataStructures = "864edb3b-99cc-5e75-8d2d-829cb0a9cfe8"
 DiffEqBase = "2b5f629d-d688-5b77-993f-72d75c75574e"
 FunctionWrappers = "069b7b12-0de2-55c6-9aab-29f3d0a68a2e"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+Parameters = "d96e819e-fc66-5662-9728-84c9c7592b0a"
```
Comment (isaacsas) [inline Project.toml:13]: Why do we need a Parameters dependency?

**Pull request closed without merging**